File: tools/sync-sitemap.mjs
/**
 * Zet de <lastmod> in sitemap.xml gelijk aan de "dateModified" uit de JSON-LD
 * van de bijbehorende pagina.
 *
 * Waarom: Google vergelijkt beide signalen. Zegt de sitemap 24 juli en de
 * pagina zelf 19 juli, dan gaat Google lastmod voor de hele site negeren.
 * De pagina is de bron van waarheid, de sitemap volgt.
 *
 * Draaien vanuit de projectmap:
 *   node tools/sync-sitemap.mjs
 *   node tools/sync-sitemap.mjs --check     (alleen melden, niets wijzigen)
 *
 * Handig na elke inhoudelijke wijziging: eerst dateModified op de pagina
 * bijwerken, dan dit script draaien.
 */

import { readFile, writeFile, access } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const SITEMAP = path.join(ROOT, 'sitemap.xml');
const BASE = 'https://startmetpokemon.nl';
const CHECK = process.argv.includes('--check');

/** Zet een sitemap-URL om naar het bestand dat hem serveert. */
function urlToFile(loc) {
  let p = loc.replace(BASE, '').replace(/^\//, '');
  if (p === '') return 'index.html';
  if (p.endsWith('/')) return p + 'index.html';
  if (!path.extname(p)) return p + '.html';
  return p;
}

/** Haal dateModified uit de JSON-LD; val terug op datePublished. */
async function pageDate(rel) {
  const full = path.join(ROOT, rel);
  try {
    await access(full);
  } catch {
    return null;
  }
  const html = await readFile(full, 'utf8');
  for (const key of ['dateModified', 'datePublished']) {
    const hits = [...html.matchAll(new RegExp(`"${key}"\\s*:\\s*"(\\d{4}-\\d{2}-\\d{2})`, 'g'))].map(
      (m) => m[1]
    );
    if (hits.length) return hits.sort().at(-1);
  }
  return null;
}

const xml = await readFile(SITEMAP, 'utf8');
const blocks = [...xml.matchAll(/<url>[\s\S]*?<\/url>/g)];

let out = xml;
const changed = [];
const missing = [];
let same = 0;

for (const m of blocks) {
  const block = m[0];
  const loc = block.match(/<loc>(.*?)<\/loc>/)[1];
  const rel = urlToFile(loc);
  const date = await pageDate(rel);

  if (!date) {
    missing.push([loc, rel]);
    continue;
  }
  const cur = block.match(/<lastmod>(.*?)<\/lastmod>/);
  if (cur && cur[1] === date) {
    same++;
    continue;
  }

  const updated = cur
    ? block.replace(cur[0], `<lastmod>${date}</lastmod>`)
    : block.replace('</loc>', `</loc>\n    <lastmod>${date}</lastmod>`);
  out = out.replace(block, updated);
  changed.push([loc, cur ? cur[1] : '-', date]);
}

for (const [loc, rel] of missing) console.log(`GEEN DATUM  ${loc}  (${rel})`);
for (const [loc, oud, nieuw] of changed) console.log(`BIJWERKEN   ${loc}  ${oud} -> ${nieuw}`);

console.log(
  `\n${same} ongewijzigd, ${changed.length} bij te werken, ${missing.length} zonder datum`
);

if (CHECK) {
  console.log('(--check: sitemap.xml niet aangepast)');
  process.exit(changed.length || missing.length ? 1 : 0);
}

if (changed.length) {
  await writeFile(SITEMAP, out, 'utf8');
  console.log('sitemap.xml bijgewerkt.');
} else {
  console.log('Niets te doen.');
}

File: tools/sync-sitemap.py
#!/usr/bin/env python3
"""
Zet de <lastmod> in sitemap.xml gelijk aan de "dateModified" uit de JSON-LD
van de bijbehorende pagina.

Waarom: Google vergelijkt beide signalen. Als de sitemap 24 juli zegt en de
pagina zelf 19 juli, gaat Google lastmod negeren voor de hele site. De pagina
is de bron van waarheid, de sitemap volgt.

Gebruik vanuit de projectmap:  python tools/sync-sitemap.py
Alleen controleren, niets wijzigen:  python tools/sync-sitemap.py --check
"""

import os
import re
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SITEMAP = os.path.join(ROOT, "sitemap.xml")
BASE = "https://startmetpokemon.nl"


def url_to_file(loc):
    """Zet een sitemap-URL om naar het bestandspad dat hem serveert."""
    path = loc.replace(BASE, "").lstrip("/")
    if path == "":
        return "index.html"
    if path.endswith("/"):
        return path + "index.html"
    if not os.path.splitext(path)[1]:
        return path + ".html"
    return path


def page_date(rel_path):
    """Haal dateModified uit de JSON-LD; val terug op datePublished."""
    full = os.path.join(ROOT, rel_path)
    if not os.path.isfile(full):
        return None
    with open(full, encoding="utf-8", errors="replace") as fh:
        html = fh.read()
    for key in ("dateModified", "datePublished"):
        found = re.findall(r'"%s"\s*:\s*"(\d{4}-\d{2}-\d{2})' % key, html)
        if found:
            return max(found)
    return None


def main():
    check_only = "--check" in sys.argv

    with open(SITEMAP, encoding="utf-8") as fh:
        xml = fh.read()

    changed, missing, skipped = [], [], []

    def replace_block(match):
        block = match.group(0)
        loc = re.search(r"<loc>(.*?)</loc>", block).group(1)
        rel = url_to_file(loc)
        date = page_date(rel)
        if date is None:
            missing.append((loc, rel))
            return block
        current = re.search(r"<lastmod>(.*?)</lastmod>", block)
        if current and current.group(1) == date:
            skipped.append(loc)
            return block
        changed.append((loc, current.group(1) if current else "-", date))
        if current:
            return block.replace(current.group(0), "<lastmod>%s</lastmod>" % date)
        return block.replace(
            "</loc>", "</loc>\n    <lastmod>%s</lastmod>" % date
        )

    new_xml = re.sub(r"<url>.*?</url>", replace_block, xml, flags=re.S)

    for loc, rel in missing:
        print("GEEN DATUM  %s  (%s)" % (loc, rel))
    for loc, old, new in changed:
        print("BIJWERKEN   %s  %s -> %s" % (loc, old, new))

    print(
        "\n%d ongewijzigd, %d bij te werken, %d zonder datum"
        % (len(skipped), len(changed), len(missing))
    )

    if check_only:
        print("(--check: sitemap.xml niet aangepast)")
        return 1 if changed or missing else 0

    if changed:
        with open(SITEMAP, "w", encoding="utf-8", newline="\n") as fh:
            fh.write(new_xml)
        print("sitemap.xml bijgewerkt.")
    else:
        print("Niets te doen.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
